src/common/StackStringStream.h

```diff
@@ -0,0 +1,169 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) 2018 Red Hat, Inc.
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation.	See file COPYING.
+ *
+ */
+
+#ifndef COMMON_STACKSTRINGSTREAM_H
+#define COMMON_STACKSTRINGSTREAM_H
+
+#include <boost/container/small_vector.hpp>
+
+#include <algorithm>
+#include <cstring>
+#include <iostream>
+#include <ostream>
+#include <sstream>
+#include <string_view>
+#include <vector>
+
+template<std::size_t SIZE>
+class StackStringBuf : public std::basic_streambuf<char>
+{
+public:
+  StackStringBuf() = default;
+  ~StackStringBuf() override = default;
+  StackStringBuf(const StackStringBuf&) = delete;
+  StackStringBuf& operator=(const StackStringBuf&) = delete;
+  StackStringBuf(StackStringBuf&& o)
+  : basic_streambuf<char>(std::move(o)),
+    vec(std::move(o.vec))
+  {
+    o.clear();
+  }
+
+  StackStringBuf& operator=(StackStringBuf&& o)
+  {
+    basic_streambuf<char>::operator=(std::move(o));
+    vec = std::move(o.vec);
+    o.clear();
+    return *this;
+  }
+
+  void push(std::string_view sv)
+  {
+    vec.reserve(vec.size() + sv.size());
+    vec.insert(vec.end(), sv.begin(), sv.end());
+  }
+
+  void clear()
+  {
+    vec.clear();
+  }
+
+  std::string_view strv() const
+  {
+    return std::string_view(vec.data(), vec.size());
+  }
+
+protected:
+  std::streamsize xsputn(const char *s, std::streamsize n)
+  {
+    push(std::string_view(s, n));
+    return n;
+  }
+
+  int overflow(int c)
+  {
+    if (traits_type::not_eof(c)) {
+      char str = traits_type::to_char_type(c);
+      push(std::string_view(&str, 1));
+      return c;
+    }
+    return EOF;
+  }
+
+private:
+
+  boost::container::small_vector<char, SIZE> vec;
+};
+
+class StackStringStream : public std::ostream
+{
+public:
+  StackStringStream() : std::ostream(&ssb) {}
+  StackStringStream(const StackStringStream& o) = delete;
+  StackStringStream& operator=(const StackStringStream& o) = delete;
+  StackStringStream(StackStringStream&& o)
+  : std::ostream(std::move(o)), ssb(std::move(o.ssb))
+  {}
+  StackStringStream& operator=(StackStringStream&& o)
+  {
+    std::ostream::operator=(std::move(o));
+    ssb = std::move(o.ssb);
+    return *this;
+  }
+  ~StackStringStream() override = default;
+
+  void clear() {
+    std::ostream::clear();
+    ssb.clear();
+  }
+
+  std::string_view strv() const {
+    return ssb.strv();
+  }
+
+private:
+  StackStringBuf<4096> ssb;
```

Comment:
this was 80 previously - is there a reason we need to go all the way up to 4k? can we test with smaller values like 256/512/1024 and see if the mds workload can still avoid most of the allocations?

Reply:
Since only a dozen or so of these is created here in the general case, I think we should leave it at a higher value. Your interest should be directed at the `ConcreteEntry` which could be smaller and would shrink the ring buffer and vectors `Log` maintains.

If we used 1024, we'd have `1088*10000` = `10880000` = 10MB in use by the ring buffer and `2*1088*100` = `2*108800` = 2*106KB for the `m_new` and `m_flush` vectors. I think that'd cover most entries IME.